File: Editorial/590_N_ary_Tree_Postorder_Traversal/Java/3_Iterative.java
class Solution {

    public List<Integer> postorder(Node root) {
        List<Integer> result = new ArrayList<>();

        // If the root is null, return the empty list
        if (root == null) return result;

        Stack<Node> nodeStack = new Stack<>(); // Stack for traversal
        Stack<Node> reverseStack = new Stack<>(); // Stack to reverse the order

        nodeStack.push(root);

        // Traverse the tree using the nodeStack
        while (!nodeStack.isEmpty()) {
            Node currentNode = nodeStack.pop();
            reverseStack.push(currentNode);

            // Push all the children of the current node to nodeStack
            for (Node child : currentNode.children) {
                nodeStack.push(child);
            }
        }

        // Pop nodes from reverseStack and add their values to the result list
        while (!reverseStack.isEmpty()) {
            Node currentNode = reverseStack.pop();
            result.add(currentNode.val);
        }

        return result;
    }
}

File: Editorial/590_N_ary_Tree_Postorder_Traversal/CPP/3_Iterative.cpp
class Solution {
public:
    vector<int> postorder(Node* root) {
        vector<int> result;

        // If the root is nullptr, return the empty vector
        if (root == nullptr) return result;

        stack<Node*> nodeStack;     // Stack for traversal
        stack<Node*> reverseStack;  // Stack to reverse the order

        nodeStack.push(root);

        // Traverse the tree using the nodeStack
        while (!nodeStack.empty()) {
            Node* currentNode = nodeStack.top();
            nodeStack.pop();
            reverseStack.push(currentNode);

            // Push all the children of the current node to nodeStack
            for (Node* child : currentNode->children) {
                nodeStack.push(child);
            }
        }

        // Pop nodes from reverseStack and add their values to the result vector
        while (!reverseStack.empty()) {
            Node* currentNode = reverseStack.top();
            reverseStack.pop();
            result.push_back(currentNode->val);
        }

        return result;
    }
};
